src/app/application/services/data-source/CallableEventSource.ts

```diff
@@ -0,0 +1,54 @@
+export const isClosed = (source: { readyState: number }) => source.readyState === 2
+export default class CallableEventSource extends EventTarget {
+  url = ''
+  withCredentials = false
+  readonly CONNECTING = 0
+  readonly OPEN = 1
+  readonly CLOSED = 2
+
+  onerror = null
+  onmessage = null
+  onopen = null
+  readyState = 2
+
+  constructor(
+    protected source: () => AsyncGenerator,
+    _configuration = {},
+  ) {
+    super()
+    this._open()
+  }
+
+  _open() {
+    (async function (self) {
+      try {
+        self.readyState = 0
+        const source = self.source()
+        for await (const res of source) {
+          self.dispatchEvent(new MessageEvent('message', {
+            data: res,
+          }))
+          if (self.readyState === 2) {
+            break
+          }
+        }
+        self.readyState = 2
+      } catch (e) {
+        console.error(e)
+        self.dispatchEvent(new ErrorEvent('error', {
+          error: e,
+        }))
+      }
+    })(this)
```

Comment:
**Question**: Why is this wrapped in an IIFE? As far as I can tell this would work just the same without it.

Reply:
I don't want `_open` to return anything and I need `async`.

Reply:
Why?

Reply:
Because I don't want anyone thinking they need to `await` it.

Reply:
> Just because something is a promise doesn’t mean it must be awaited for use

No but somebody might think that when their editor tells them it returns a promise. This is a potential negative consequence, and I can't think of any potential negative consequences of leaving as is. If you can think of one let me know and I'll gladly change it up, otherwise it feels like opinion.

If a negative consequence arises at a later date we can also change.

Reply:
Resolving this, I don't think its a major thing but please let me know if I should reopen